providers/common/ciphers/aes.c

```diff
@@ -0,0 +1,424 @@
+{
+    if (iv != NULL && ctx->mode != EVP_CIPH_ECB_MODE) {
+        if (ivlen != AES_BLOCK_SIZE)
+            return 0;
+        memcpy(ctx->iv, iv, AES_BLOCK_SIZE);
+    }
+    ctx->enc = enc;
+
+    return 1;
+}
+
+static int aes_einit(void *vctx, const unsigned char *key, size_t keylen,
+                           const unsigned char *iv, size_t ivlen)
+{
+    PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
+
+    if (!PROV_AES_KEY_generic_init(ctx, iv, ivlen, 1))
+        return 0;
+    if (key != NULL) {
+        if (keylen != ctx->keylen)
+            return 0;
+        return ctx->ciph->init(ctx, key, ctx->keylen);
+    }
+
+    return 1;
+}
+
+static int aes_dinit(void *vctx, const unsigned char *key, size_t keylen,
+                     const unsigned char *iv, size_t ivlen)
+{
+    PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
+
+    if (!PROV_AES_KEY_generic_init(ctx, iv, ivlen, 0))
+        return 0;
+    if (key != NULL) {
+        if (keylen != ctx->keylen)
+            return 0;
+        return ctx->ciph->init(ctx, key, ctx->keylen);
+    }
+
+    return 1;
+}
+
+static int aes_block_update(void *vctx, unsigned char *out, size_t *outl,
+                            size_t outsize, const unsigned char *in, size_t inl)
+{
+    PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
+    size_t nextblocks = fillblock(ctx->buf, &ctx->bufsz, AES_BLOCK_SIZE, &in,
+                                  &inl);
+    size_t outlint = 0;
+
+    /*
+     * If we're decrypting and we end an update on a block boundary we hold
+     * the last block back in case this is the last update call and the last
+     * block is padded.
+     */
+    if (ctx->bufsz == AES_BLOCK_SIZE
+            && (ctx->enc || inl > 0 || !ctx->pad)) {
+        if (outsize < AES_BLOCK_SIZE)
+            return 0;
+        if (!ctx->ciph->cipher(ctx, out, ctx->buf, AES_BLOCK_SIZE))
+            return 0;
+        ctx->bufsz = 0;
+        outlint = AES_BLOCK_SIZE;
+        out += AES_BLOCK_SIZE;
+    }
+    if (nextblocks > 0) {
+        if (!ctx->enc && ctx->pad && nextblocks == inl) {
+            if (!ossl_assert(inl >= AES_BLOCK_SIZE))
+                return 0;
+            nextblocks -= AES_BLOCK_SIZE;
+        }
+        outlint += nextblocks;
+        if (outsize < outlint)
+            return 0;
+        if (!ctx->ciph->cipher(ctx, out, in, nextblocks))
+            return 0;
+        in += nextblocks;
+        inl -= nextblocks;
+    }
+    if (!trailingdata(ctx->buf, &ctx->bufsz, AES_BLOCK_SIZE, &in, &inl))
+        return 0;
+
+    *outl = outlint;
+    return inl == 0;
+}
+
+static int aes_block_final(void *vctx, unsigned char *out, size_t *outl,
+                           size_t outsize)
+{
+    PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
+
+    if (ctx->enc) {
+        if (ctx->pad) {
+            padblock(ctx->buf, &ctx->bufsz, AES_BLOCK_SIZE);
+        } else if(ctx->bufsz == 0) {
+            *outl = 0;
+            return 1;
+        } else if(ctx->bufsz != AES_BLOCK_SIZE) {
```

Comment:
Space after the if here and the one above.

Reply:
Fixed.